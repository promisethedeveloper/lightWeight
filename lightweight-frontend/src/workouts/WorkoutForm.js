import React, { useState, useContext } from "react";
import { useHistory } from "react-router-dom";
import UserContext from "../auth/UserContext";
import Alert from "../common/Alert";
import LightWeightAPI from "../api/api";

const WorkoutForm = () => {
	const INITIALSTATE = {
		dayOfWeek: "Monday",
		description: "",
	};

	const { currentUser } = useContext(UserContext);
	const { id: user_id } = currentUser;

	const history = useHistory();
	const [formData, setFormData] = useState(INITIALSTATE);

	const [formErrors, setFormErrors] = useState([]);

	/**
	 * Handle form submit:
	 *
	 * Calls login function prop and if successful, redirect to "/workouts"
	 */

	const handleSubmit = async (e) => {
		e.preventDefault();
		try {
			await LightWeightAPI.createWorkout(formData);
			history.push("/workouts");
			setFormData(INITIALSTATE);
		} catch (error) {
			setFormErrors(error);
		}
	};

	const handleChange = (e) => {
		const { name, value } = e.target;
		setFormData((data) => ({ ...data, user_id: user_id, [name]: value }));
	};

	return (
		<div className="LoginForm">
			<div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
				<h3 className="mb-3">Add workout day</h3>

				<div className="card">
					<div className="card-body">
						<form onSubmit={handleSubmit}>
							<div className="form-group">
								<label htmlFor="dayOfWeek">Day of week</label>
								<select
									name="dayOfWeek"
									onChange={handleChange}
									className="form-control"
									value={formData.dayOfWeek}
									id="dayOfWeek"
								>
									<option value="Monday">Monday</option>
									<option value="Tuesday">Tuesday</option>
									<option value="Wednesday">Wednesday</option>
									<option value="Thursday">Thursday</option>
									<option value="Friday">Friday</option>
									<option value="Saturday">Saturday</option>
									<option value="Sunday">Sunday</option>
								</select>
							</div>
							<div className="form-group">
								<label htmlFor="description">Description</label>
								<input
									id="description"
									name="description"
									className="form-control"
									value={formData.description}
									onChange={handleChange}
									placeholder="ex. Chest day"
									required
								/>
							</div>
							{formErrors.length ? (
								<Alert type="danger" messages={formErrors} />
							) : null}
							<button
								className="btn float-right font-weight-bold"
								style={{ backgroundColor: "#461111", color: "white" }}
							>
								Save
							</button>

							<a
								className="btn float-right font-weight-bold"
								style={{
									backgroundColor: "#FF1E00",
									color: "white",
									marginRight: "5px",
								}}
								href="/workouts"
								role="button"
							>
								Cancel
							</a>
						</form>
					</div>
				</div>
			</div>
		</div>
	);
};

export default WorkoutForm;
