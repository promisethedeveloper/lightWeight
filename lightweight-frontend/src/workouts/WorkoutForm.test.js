import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import WorkoutForm from "./WorkoutForm";
import UserContext from "../auth/UserContext";

/** Smoke test */
it("should render without crashing", () => {
	const currentUser = { username: "John" };
	render(
		<MemoryRouter>
			<UserContext.Provider value={{ currentUser }}>
				<WorkoutForm />
			</UserContext.Provider>
		</MemoryRouter>
	);
});

/** Snapshot test */
it("should match snapshots", () => {
	const currentUser = { username: "John" };

	const { asFragment } = render(
		<MemoryRouter>
			<UserContext.Provider value={{ currentUser }}>
				<WorkoutForm />
			</UserContext.Provider>
		</MemoryRouter>
	);
	expect(asFragment()).toMatchSnapshot();
});

it("should render form label text", () => {
	let currentUser = { username: "Promise" };
	const { getByLabelText } = render(
		<MemoryRouter>
			<UserContext.Provider value={{ currentUser }}>
				<WorkoutForm />
			</UserContext.Provider>
		</MemoryRouter>
	);
	expect(getByLabelText("Day of week")).toBeInTheDocument();
	expect(getByLabelText("Description")).toBeInTheDocument();
});

it("should render Save and Cancel buttons", () => {
	let currentUser = { username: "Promise" };
	const { getByText } = render(
		<MemoryRouter>
			<UserContext.Provider value={{ currentUser }}>
				<WorkoutForm />
			</UserContext.Provider>
		</MemoryRouter>
	);
	expect(getByText("Save")).toBeInTheDocument();
	expect(getByText("Cancel")).toBeInTheDocument();
});
