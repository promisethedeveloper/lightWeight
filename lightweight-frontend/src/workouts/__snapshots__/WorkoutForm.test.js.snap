// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshots 1`] = `
<DocumentFragment>
  <div
    class="LoginForm"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h3
        class="mb-3"
      >
        Add workout day
      </h3>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label
                for="dayOfWeek"
              >
                Day of week
              </label>
              <select
                class="form-control"
                id="dayOfWeek"
                name="dayOfWeek"
              >
                <option
                  value="Monday"
                >
                  Monday
                </option>
                <option
                  value="Tuesday"
                >
                  Tuesday
                </option>
                <option
                  value="Wednesday"
                >
                  Wednesday
                </option>
                <option
                  value="Thursday"
                >
                  Thursday
                </option>
                <option
                  value="Friday"
                >
                  Friday
                </option>
                <option
                  value="Saturday"
                >
                  Saturday
                </option>
                <option
                  value="Sunday"
                >
                  Sunday
                </option>
              </select>
            </div>
            <div
              class="form-group"
            >
              <label
                for="description"
              >
                Description
              </label>
              <input
                class="form-control"
                id="description"
                name="description"
                placeholder="ex. Chest day"
                required=""
                value=""
              />
            </div>
            <button
              class="btn float-right font-weight-bold"
              style="background-color: rgb(70, 17, 17); color: white;"
            >
              Save
            </button>
            <a
              class="btn float-right font-weight-bold"
              href="/workouts"
              role="button"
              style="background-color: rgb(255, 30, 0); color: white; margin-right: 5px;"
            >
              Cancel
            </a>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
